com.unity.sg2/Editor/GraphUI/EditorCommon/Preview/PreviewManager.cs

```diff
@@ -101,7 +99,142 @@ internal void Initialize(ShaderGraphModel graphModel, MainPreviewView mainPrevie
 
         static bool IsMainContextNode(IGraphElementModel nodeModel)
         {
-            return nodeModel is GraphDataContextNodeModel contextNode && contextNode.graphDataName == new Defs.ShaderGraphContext().GetRegistryKey().Name;
+            return nodeModel is GraphDataContextNodeModel contextNode
+                    && contextNode.graphDataName == new Defs.ShaderGraphContext().GetRegistryKey().Name;
+        }
+
+        internal void UpdateReferencesAfterUndoRedo(
+            GraphModelStateComponent graphModelStateComponent,
+            ShaderGraphModel graphModel)
+        {
+            m_GraphModelStateComponent = graphModelStateComponent;
+            m_GraphModel = graphModel;
+
+            m_PreviewHandlerInstance.SetActiveGraph(m_GraphModel.GraphHandler);
+        }
+
+        /// <summary>
+        ///  Called to reinitialize the textures for the previews when the graph is saved
+        /// </summary>
+        internal void HandleGraphReload(
+            GraphModelStateComponent graphModelStateComponent,
+            ShaderGraphModel graphModel,
+            MainPreviewView mainPreviewView)
+        {
+            m_GraphModelStateComponent = graphModelStateComponent;
+            m_GraphModel = graphModel;
+            m_MainPreviewView = mainPreviewView;
+
+            ReassignAllNodePreviewTextures();
+        }
+
+        /// <summary>
+        /// After a save or an undo/redo, all the GTF node views get recreated so we gotta go through and reassign any cached node preview textures
+        /// </summary>
+        internal void ReassignAllNodePreviewTextures()
+        {
+            using (var stateUpdater = m_GraphModelStateComponent.UpdateScope)
+            {
+                foreach (var (nodeName, nodeGuid) in m_NodeLookupTable)
+                {
+                    m_GraphModel.TryGetModelFromGuid(nodeGuid, out var nodeModel);
+                    // Don't update the node right away if it is dirty
+                    if (nodeModel is GraphDataNodeModel graphDataNodeModel
+                        && !m_PreviewHandlerInstance.IsNodeDirty(graphDataNodeModel.graphDataName))
+                    {
+                        m_PreviewHandlerInstance.RequestNodePreviewTexture(nodeName, out var nodeRenderOutput, out var shaderMessages, graphDataNodeModel.NodePreviewMode);
+                        graphDataNodeModel.OnPreviewTextureUpdated(nodeRenderOutput);
+
+                        stateUpdater.MarkChanged(graphDataNodeModel);
+                    }
+                }
+            }
+        }
```

Comment:
Why don't the node UIs just request their own preview textures during their initialization?